### ISSUE TITLE
No keyboard events on REAPER/Linux (VST2)

### ISSUE THREAD

**Issue by lucianoiam** [DISTRHO/DPF#298]
The VST2 callback for plugin keyboard events just never gets called, here:

https://github.com/DISTRHO/DPF/blob/8401d865cd9fce9edbeb9c653420bc0f8a39358c/distrho/src/DistrhoPluginVST.cpp#L663

Not sure where to further look

- ReaTune VST2 has some text input boxes and it works (maybe not a good quality example)

- Made sure e9563d1 is present anyways the code in there depends on this.

- On REAPER/Mac callback fires but only after toggling the UI by clicking this button twice:

<img width="150" alt="Screen Shot 2021-07-06 at 01 37 34" src="https://user-images.githubusercontent.com/930494/124524270-c0bb6080-ddfa-11eb-8db0-14143406b2ba.png">

1. Insert plugin. No effEditKeyDown.
2. Hide custom UI
3. Show custom UI
4. Keyboard works

The trick does not have any effect on Linux

**Comment by falkTX** [comment]
I noticed this too, mostly an issue on macOS where we cannot make the plugin UI receive proper keyboard focus.
I tried several things yesterday, but could find out exactly what is wrong.

If you want something to try, check with the standalone/jack executable.
At this point it won't receive keyboard focus no matter what we do. My guess is that fixing this keyboard focus on the standalone will result on it working for the lv2/vst2 plugin side too.

**Comment by falkTX** [comment]
JUCE has a flag for plugins to indicate they want keyboard focus, might be worth investigating what it does.

**Comment by lucianoiam** [comment]
Saw some old lengthy discussion about keyboard input in both REAPER and JUCE forums; it clearly seems a hard nut to crack and not particular to DPF. This needs digging deep in code as suggested.

Started by trying to determine what is the expected behavior.

I realized REAPER won't move focus to the plugin right after instantiating it.  When for example "ReaCast" is added, the focus is retained in the list to the left and pressing keys results in beep.

<img width="810" alt="Screen Shot 2021-07-06 at 14 06 21" src="https://user-images.githubusercontent.com/930494/124597153-69050f80-de63-11eb-832d-42f6584bc12d.png">

Clicking on an text input field moves focus as expected:

<img width="804" alt="Screen Shot 2021-07-06 at 14 06 34" src="https://user-images.githubusercontent.com/930494/124597159-6b676980-de63-11eb-85ad-f27dac5a3d84.png">

**Getting beeps for keystrokes right after adding a DPF plugin is actually fine.**

The issue is that currently it does not seem possible for the user to request focus on a `PuglWrapperView`. Adding a line in mac.m does the job:

```Objective-C
- (void)mouseDown:(NSEvent*)event
{
  [self.window makeFirstResponder:self];  // <--
  const NSPoint         wloc = [self eventLocation:event];
  const NSPoint         rloc = [NSEvent mouseLocation];
  ...
```

Basically we tell that view to become the first responder when the user clicks on it. Now after clicking on the plugin view, `effEditKeyDown` events start to trigger normally when pressing keys and the beep is gone. 

*Wild assumption: REAPER will not offer VST2 key events to the plugin unless its view of any of its subviews is the first responder, aka Mac keyboard focus.*

Pugl view actually asks to become first responder on instantiation but then REAPER takes the focus back. I first discovered that by stitching a NSTextView to the native view and calling `makeFirstResponder:` during init (no effect) and calling it 0.5s later (works -- input gets focused). It is not crazy to think this is by design to prevent plugins from stealing focus when inserting them.

It is probably cleaner to patch this within DPF at `Window::PrivateData::onPuglMouse()`

```C++
void Window::PrivateData::onPuglMouse(const Widget::MouseEvent& ev)
{
    DGL_DBGp("onPuglMouse : %i %i %f %f\n", ev.button, ev.press, ev.pos.getX(), ev.pos.getY());

    if (ev.press) {
        focus();
    }
    ...
```

A last solution is to not change anything and leave it up to the plugin developer. They should remember that 1) catching keystrokes is not possible unless `DGL::Window::focus()` is called first [ which in turn calls `puglGrabFocus()` and lastly `makeFirstResponder:` ] and 2) asking for focus on init does not work because REAPER will take focus back when adding the plugin.

Regarding Linux still no idea.

**Comment by falkTX** [comment]
Is that also the case for standalones? What about LV2?

I find it weird that Linux has similar behaviour, I expect focus to be given to the plugin window without issues.
On Linux, other hosts do not have such quirks, correct?

In any case, this is a good find.
Maybe worth trying to give the UI focus on the pugl "window appeared on screen" event.

**Comment by lucianoiam** [comment]
It actually seems it is just REAPER regardless the platform and this is by design. My bet after the NSTextView experiment is that REAPER reclaims focus right after plugin initialization so the focus is not removed from the shiny plugin list on the left. The real answer comes by sniffing events on every platform but that takes more time so for now it is an assumption.

So far other hosts have not exhibited this behavior on any platform. I regularly test stuff on Carla, Bitwig, REAPER and Live on the three OSes.

To me DPF is correctly doing its job and REAPER the "offender". If we requested focus on a "window appeared on screen" event, then we are kind of trying to win a focus stealing race :)  [ 1) plugin init 2) host reclaims 3) plugin reclaims ]. Shouldn't the focus be moved only by explicit user request?. It is the "keyboard focus" after all. Plugins implementing knobs and sliders and no key shortcuts don't need it, in such cases it seems better to let the host decide  (in the case of REAPER, to keep the list of inserts focused). Then if the plugin needs keystrokes, whether for global shortcuts (rare) or text boxes, it can call focus() when needed. In the case of a text box, focus() would be naturally called when the text box is clicked.

Now the "window appeared on screen" event is mentioned, that is not exposed in DGL right? I recall searching for something like that in the past for an unrelated task.

I will prepare a test modifying d_parameters to print onKeyboard() events, with and without the proposed patch, VST2/LV2, all platforms, all hosts. So we get a conclusive answer on the focus thing, independently on how to gain it. Need a little time.

**Comment by falkTX** [comment]
> Now the "window appeared on screen" event is mentioned, that is not exposed in DGL right? I recall searching for something like that in the past for an unrelated task.

Correct, the event is in each separate file in pugl.

I mainly ask if this would fix anything because right now DPF standalones on macOS do not automatically get focused on start, which is very annoying.
I actually can't give them keyboard focus when running from CLI.
Will try with your idea later on.



> I will prepare a test modifying d_parameters to print onKeyboard() events, with and without the proposed patch, VST2/LV2, all platforms, all hosts. So we get a conclusive answer on the focus thing, independently on how to gain it. Need a little time.

No need for this, if you clone and place imgui on the same directory as DPF, you can build the ImguiSimpleGain example plugin, which includes a tiny text editor field.
(for now build is manual, and only through makefile, but it is possible to make it work)

![Screenshot_20210706_154502](https://user-images.githubusercontent.com/1334853/124620157-23e7da00-de71-11eb-804c-4cd678e78f1b.png)

**Comment by lucianoiam** [comment]
> I mainly ask if this would fix anything because right now DPF standalones on macOS do not automatically get focused on start, which is very annoying.
> I actually can't give them keyboard focus when running from CLI.

This is a macOS specific issue and (I think) comes from the fact that the standalone version is lacking some details to be a "real" Cocoa app. For example I see no `NSApplicationMain()` call or app bundle .plist defined. A quick fix involves adding a line to pugl:

```Objective-C
PuglWorldInternals*
puglInitWorldInternals(PuglWorldType type, PuglWorldFlags PUGL_UNUSED(flags))
{
  PuglWorldInternals* impl =
    (PuglWorldInternals*)calloc(1, sizeof(PuglWorldInternals));

  impl->app = [NSApplication sharedApplication];

  if (type == PUGL_PROGRAM) {
    [impl->app setActivationPolicy:NSApplicationActivationPolicyRegular];  // <-- this is new
    impl->autoreleasePool = [NSAutoreleasePool new];
  }

  return impl;
}
```

Note some details still don't work completely well, like the menu bar not being focusable at first or Cmd+Q not working for quitting the app. Probably the real solution is to provide app bundles for the standalone, just like it is done for MacVST, and define some little metadata there.

> if you clone and place imgui on the same directory as DPF, you can build the ImguiSimpleGain example plugin, which includes a tiny text editor field.
> (for now build is manual, and only through makefile, but it is possible to make it work)
> 

That's great, will have a look at it.

**Comment by lucianoiam** [comment]
FYI here are some test results after patching ImGuiUI.cpp to grab focus on click. They are consistent with previous findings, and somewhat surprising on Windows.

```C++
bool ImGuiUI::onMouse(const MouseEvent& event)
{
    if (event.press) {
        getWindow().focus();
    }
...
```

Result types:

A. Text input works without patch
B. Text input works only after applying patch
X. Could not test due to UI not properly showing  (known LV2 issue #294)

Linux:  REAPER/VST (B) — REAPER/LV2 (X) —  Carla.git/VST (A) — Carla.git/LV2 (A) — Bitwig/VST (A)
Mac:  REAPER/VST (B) — Live/VST (A)  —  Carla.kx/VST (B **)  — Carla.kx/LV2  (X)
Windows   REAPER/VST (A)  — Live/VST (A)  — Carla.kx/VST (B)

** View flickers, UI feels a bit unresponsive


Is the test below what you propose? I could not get it working on REAPER/Mac, PUGL_MAP just never fires. 

WindowPrivateData.cpp

```C++
    ///< View made visible, a #PuglEventMap
    case PUGL_MAP:
        puglGrabFocus(view);    // <-- new line
        break;
```

These are the first pugl events when adding a plugin:

```
pugl event: Focus in 0      ★
pugl event: Focus out 0     ★
pugl event: Create
pugl event: Destroy
pugl event: Create
pugl event: Configure    0.0    0.0  600.0  400.0
pugl event: Expose       0.0    0.0  600.0  400.0
pugl event: Expose       0.0    0.0  600.0  400.0
pugl event: Expose       0.0    0.0  600.0  400.0
pugl event: Expose       0.0    0.0  600.0  400.0
```

**Comment by falkTX** [comment]
Coming back to this, did anything change that made this behave better?
We now have the VST2 key up/down events handled on plugin side, so maybe that is all we need?

Carla from latest git should also send these values if using windows or macOS (or linux it gives the plugin UI focus, so that is not needed)

**Comment by lucianoiam** [comment]
Well I moved away from DGL/Pugl but even for ExternalWindow some brute force is needed to have the plugin window keep the keyboard focus for itself. This is true for REAPER and Bitwig Studio but not for Carla. I think you mentioned some hosts go to great extents to keep focus for themselves and my own observation goes on par with that.

So I opted to bypass the host and fix this at the windowing system level. For the GTK based webview the workaround consisted in calling `gdk_seat_grab()`/`gdk_seat_ungrab` with `GDK_SEAT_CAPABILITY_KEYBOARD` when the keyboard is needed.
